### DISCUSSION TITLE
Disable consult buffer preview for remote buffers

### DISCUSSION THREAD

**Discussion by mgraham** [minad/consult#969] [Q&A] [answered]
I have been trying to disable consult buffer preview for remote buffers, but I am unable to get the solution in https://github.com/minad/consult/issues/224 and the [wiki](https://github.com/minad/consult/wiki#do-not-preview-exwm-windows-or-tramp-buffers) to work.

I'm tracing through the code in the solution and it looks to me that it sets `filter` to a lambda that returns either `cand` (the candidate) or `nil` (which it does when the candidate is a remote file):

```
(defun consult-buffer-state-no-tramp ()
  "Buffer state function that doesn't preview Tramp buffers."
  (let ((orig-state (consult--buffer-state))
        (filter (lambda (action cand)
                  (if (and cand
                           (or (eq action 'return)
                               (let ((buffer (get-buffer cand)))
                                 (and buffer
                                      (not (file-remote-p (buffer-local-value 'default-directory buffer)))))))
                      cand
                    nil))))
    (lambda (action cand)
      (funcall orig-state action (funcall filter action cand)))))

(setq consult--source-buffer
      (plist-put consult--source-buffer :state #'consult-buffer-state-no-tramp))
```

But even when I change the `filter` lambda so that it always returns `nil`, `consult-buffer` still gives me previews:
```
(defun consult-buffer-state-no-tramp ()
  "Buffer state function that doesn't preview any buffers."
  (let ((orig-state (consult--buffer-state))
        (filter (lambda (action cand)
                  (if (and cand
                           (or (eq action 'return)
                               (let ((buffer (get-buffer cand)))
                                 (and buffer
                                      (not (file-remote-p (buffer-local-value 'default-directory buffer)))))))
                      nil  ;; always return nil - for remote and local buffers
                    nil))))
    (lambda (action cand)
      (funcall orig-state action (funcall filter action cand)))))

(setq consult--source-buffer
      (plist-put consult--source-buffer :state #'consult-buffer-state-no-trampl))
```

I would expect that the above would disable previews for all buffers, but it doesn't.  

Simplifying it further, the following code also fails to disable previews:
```
(defun consult-buffer-state-never-preview ()
  "Buffer state function that doesn't preview any buffers."
  (let ((orig-state (consult--buffer-state)))
    (lambda (action cand)
      (funcall orig-state action nil))))

(setq consult--source-buffer
      (plist-put consult--source-buffer :state #'consult-buffer-state-never-preview))
```

I don't know enough about the consult internals to know why these don't work.  I went looking for the definition of `consult--buffer-state` but it's docs say that it is defined as a closure, not a `defun`, and I'm afraid I'm at the limit of my elisp skills!

**Comment by mgraham** [comment]
If I understand the docs for `consult--with-preview` correctly (and I'm not sure that I do), I think the idea is that if I wanted to use the state function to disable previews entirely, I would want this filter lambda to return `nil` when `action` is `'preview`, and return `cand` when action is anything else.  Essentially, we're intercepting the state function and calling `(consult--buffer-state 'preview nil)`?

I *think* that is what I am doing below:

```
(defun consult-buffer-state-disable-previews ()
  "Buffer state function that doesn't preview Tramp buffers."
  (lambda (action cand)
    (let ((orig-state (consult--buffer-state))
          (filter (lambda (action cand)
                    (if (eq action 'preview) nil
                      cand))))
      (lambda (action cand)
        (funcall (orig-state) action (funcall filter action cand))))))

(setq consult--source-buffer
      (plist-put consult--source-buffer :state #'consult-buffer-state-disable-previews))
```

However, when I run `consult-buffer` and cycle through the candidates, I do still see previews.

Note that at this point, this is nothing to do with remote buffers.  I am just trying to see if I understand how I'm supposed to write the state function.

Have I got the right idea?  That calling `consult--buffer-state` with a `cand` of `nil` when `action` is `'preview` will disable previews?  Or am I misunderstanding things?

**Comment by minad** [comment]
@mgraham Disabling buffer preview for certain buffers has been requested often. Would you like to prepare a patch which adds a variable `consult-preview-excluded-buffers`, similar to the already existing `consult-preview-excluded-files`? In the long term, this may be better than continuing with the potentially brittle wiki code.

**Comment by mgraham** [comment]
@minad - Sure - I can take a stab at that!  I will have to spend some time grokking the consult internals first.

It looks like `consult-preview-excluded-files` is a regexp.  I'm guessing `consult-preview-excluded-buffers` should probably be able to match on buffer name, major-mode, minor-mode and possibly other buffer attributes.  Maybe it should be a lambda?

**Reply by minad** [reply]
There exists the new function `buffer-match-p` which we could use. We could use it for `consult-preview-excluded-buffers` and also replace `consult-buffer-filter` with it for consistency:

~~~ elisp
(buffer-match-p buf `(or ,@consult-buffer-filter))
~~~

My only gripe with this `buffer-match-p` function is that it is not particularly efficient. So we should maybe implement our own filter, which creates an efficient predicate function and a regexp.

~~~ elisp
(setq consult-buffer-filter
  '("regexp"      ;; stringp: regexp matchig buffer name
    (var list...) ;; consp: (member (buffer-local-value var) list), e.g., var=major-mode
    predicate))   ;; functionp: predicate function
~~~

**Reply by minad** [reply]
@mgraham I looked again a little more deeply into the existing Consult code, and I think a larger refactoring/consolidation concerning buffer filtering is needed. I will work on this when I find some time. Just letting you know, such that you don't have to spent time on creating a PR.

The rough plan:

1. `consult--buffer-query` will be deprecated in favor of a new function `consult--buffer-list`.
2. A new function `consult--buffer-predicate` will take a flexible filter expression (similar to `buffer-match-p`) as argument but compiles it to an efficient predicate function. 
3. `consult--buffer-list` will take such a filter expression (or a predicate function) as argument, compiles it and returns the filtered list.

Then we can use `consult--buffer-predicate` to handle the new `consult-preview-excluded-buffers` and use `consult--buffer-list` instead of `consult--buffer-query`.

**Reply by mgraham** [reply]
Sounds good. Let me know if I can help

**Reply by minad** [reply, marked as answer]
For now I went with a simple solution and added `consult-preview-excluded-buffers` based on `buffer-match-p`. See https://github.com/minad/consult/commit/e1621f4e4be6c7078373c74346311e640e16338f.

**Reply by TristanCacqueray** [reply]
Hello @minad , could you please explain how to use consult-preview-excluded-buffers to exclude remote files? Using `(setq consult-preview-excluded-buffers 'file-remote-p)` fails with `Error in post-command-hook (consult--preview-post-command-hook): (wrong-type-argument stringp #<buffer *shell*>)`. Do we need to setup a custom predicate to get the buffer's file before using `file-remote-p` ?

**Reply by minad** [reply]
@TristanCacqueray Check out the documentation of `buffer-match-p`. I think you need a custom predicate, yes.

**Reply by TristanCacqueray** [reply]
Alright, the following predicate seems to work, though the debug statement shows a weird file-name for closed remote buffers (from recentf), it looks like they point at the currently opened buffer, but it seems like `file-remote-p` works as expected. Could this be added to the Wiki?

```elisp
(defun buffer-remote-p (buf)                                                                                                                                                                                                                                                                                      
  "Return t when BUF is remote."                                                                                                                                                                                                                                                                                
  ;; (message "HERE [%s - %s]" buf (buffer-file-name buf))                                                                                                                                                                                                                                                         
  (if-let ((fp (buffer-file-name buf)))                                                                                                                                                                                                                                                                         
      (file-remote-p fp)                                                                                                                                                                                                                                                                                        
    nil))

(setq consult-preview-excluded-buffers 'buffer-remote-p)
```

Thanks!

**Reply by JasZhe** [reply]
@minad I've noticed that `non-essential . t` is set by default in `consult-preview-variables`. This is then used in `consult--find-file-temporarily`, but I'm wondering if it would make sense to move this up maybe to `consult--temporary-files` instead since we call `expand-file-name` there which would trigger a tramp call I think from the `file-name-handler-alist`

This has come up from some discussion in https://github.com/jdtsmith/consult-jump-project/pull/5

**Reply by minad** [reply]
@JasZhe @jdtsmith Please check if https://github.com/minad/consult/commit/dea4eaf621b6477cca514bfa32c950004fb9cc99 works for you. I am not sure though if this introduces other misbehavior.

**Reply by JasZhe** [reply]
hey sorry got a little busy last night, but yeah that did fix the issue thank you so much!

**Reply by minad** [reply]
Thanks for checking. These Tramp issues are subtle since many functions are infected and it has happened many times to me that I've forgotten some calls.

**Reply by jdtsmith** [reply]
Yes, I find that too.  I wish TRAMP had a more robust method to mark some connections as "unavailable", and quickly error out.  In this era of VPNs etc., the idea that all previously known or seen TRAMP connections should be reliably to reconnect to at all times, and the only way to find out is to make the connection and hang indefinitely is unfortunate.  

I do wonder if there's a way to "fail quickly" for many of these, and mark failing connections unavailable.  I might take it up on the TRAMP list.

**Reply by minad** [reply]
Yes, bailing out more quickly sounds good. Another problem is that there is no obvious separation between pure and IO operations. It would be better if there was a set of file path operations which are guaranteed to never access the file system and a set of functions which necessarily have to access the file system. For example a while ago Tramp started to modify `abbreviate-file-name` which I had used mainly for display purposes, and which had been pure until then. Suddenly it got slow due to Tramp. I think it is acceptable to simply not abbreviate remote paths. On the other hand such a line is hard to draw and the result is that Tramp affects *everything* in Emacs. Many things just work as expected, but highly dynamic interactive functionality is hurt quite a bit (interactive completion filtering, previews, etc.). Furthermore there is always the risk that a buffer suddenly hangs, which is a result of the synchronous single threaded nature of Emacs. An alternative to Tramp would be to use an alternative file system layer (e.g. sshfs), but then process invocation would not happen on the remote side and the problem is essentially just pushed a level deeper. It is just as bad when the file system hangs.

**Reply by jdtsmith** [reply]
I just searched around the TRAMP mailing list and found a 6yo thread[ I myself created](https://lists.gnu.org/archive/html/tramp-devel/2019-11/msg00002.html) regarding inaccessible hosts.  Sigh.

**Reply by minad** [reply]
I suggest to create an Emacs bug/feature request on the Emacs issue tracker instead.
